### PR TITLE
feat(kvstore): support mamba l2 cache transfers

### PR DESCRIPTION
**Pull request by XucSh**
## Summary
This PR adds L2 cache support for Mamba cache alongside the existing KV cache L2 path.

The core change is to generalize cache movement into CacheTransferUnit, which carries the cache kind (KV or Mamba) plus source and destination slots/pages. WriteBackOperation and LoadBackOperation can now carry mixed KV/Mamba transfer units without adding Mamba-specific scheduler branches everywhere.

Key behavior:

- Mamba cache can be written back from GPU to host memory.
- Mamba cache can be loaded back from host memory to GPU.
- Mamba host memory supports eviction when host Mamba slots are exhausted.
- Mamba loadback completion is synchronized through stream fence semantics, not scheduler loadback-done events.
- Prefix reuse now requires all cache types for a token prefix to exist. KV-only cache is not considered reusable when the corresponding Mamba state is missing.
- Mamba state depth is tracked explicitly so page-aligned prefix states and non-exact retraction recovery states are not confused.

Important design points:

- CacheTransferUnit avoids duplicating writeback/loadback logic for KV and Mamba.
- TreeNode tracks device/host Mamba slots and their token depth.
- FindLastMambaNode(..., require_exact_depth=true) is used for prefix cache reuse.
- Retraction recovery may use non-exact live Mamba state, but normal prefix reuse may not.
- Host-side Mamba eviction detaches only the host Mamba state; future prefix reuse still requires KV and Mamba to both be present.

## Test Plan
`ts serve \
  --model nvidia/Qwen3.5-397B-A17B-NVFP4 \
  --host 127.0.0.1 \
  --port 12346 \
  --trust-remote-code \
  --max-model-len 120000 \
  --gpu-memory-utilization 0.95 \
  --max-num-seqs 16 \
  --max-total-tokens 1000000 \
  --chunked-prefill-size 128 \
  --max-prefill-tokens 128 \
  --kv-cache-dtype fp8_e4m3 \
  --quantization nvfp4 \
  --enable-mamba-l2 \
  --mamba-l2-host-slots 512 \
  --mamba-l2-io-backend kernel \
  --enable-cache-report \
  --moe-backend flashinfer_trtllm \
  --attention-backend trtllm \
  --attn-tp-size 4`

`evalscope eval \
  --model nvidia/Qwen3.5-397B-A17B-NVFP4 \
  --api-url http://127.0.0.12346/v1 \
  --api-key EMPTY_TOKEN \
  --datasets aime25 \
  --eval-batch-size 16 \
  --generation-config '{"max_tokens":100000}'`

mamba loadback/writeback can be seen. The score is 0.9

### CODE REVIEW

File: tokenspeed-scheduler/csrc/fsm/forward_events.cpp [hunk 1]
```diff
@@ -323,13 +334,30 @@ std::variant<Draining, Finished> FinishEvent::apply(ForwardStateT&& state) {
...
-        std::unique_ptr<HostNodeRef> host_node_ref = std::make_unique<HostNodeRef>(match.device.last_node);
+        transfers = BuildWriteBackPairs(write_diff);
+        device_node_ref = std::make_unique<DeviceNodeRef>(match.device.last_node);
+        host_node_ref = std::make_unique<HostNodeRef>(match.device.last_node);
+    } else {
+        device_node_ref = std::make_unique<DeviceNodeRef>(mamba_node);
+    }
 
-        return Draining{BuildWriteBackPairs(write_diff), std::move(device_node_ref), std::move(host_node_ref)};
+    if (need_mamba_writeback) {
```
Comment (tuanzhangCS) [inline tokenspeed-scheduler/csrc/fsm/forward_events.cpp]: Currently, we only call write_back when a request finishes or is retracted, and we only write back the last Mamba slot of that request. Should we also drain Mamba slots during the prefill stage to improve the cache hit rate?

File: tokenspeed-scheduler/csrc/fsm/forward_events.cpp [hunk 2]
```diff
@@ -439,17 +467,37 @@ Retracting ScheduleRetractEvent::applyRetract(ForwardStateT&& state) {
...
+            if (!hybrid_prefix_cache_->EnsureMambaHostCapacityByEvict(1, terminal)) {
+                throw std::logic_error("ScheduleRetractEvent: failed to allocate host slot for mamba writeback");
+            }
+            auto host_slot = hybrid_prefix_cache_->AllocateHostMamba();
+            if (host_slot == nullptr) {
+                throw std::logic_error("ScheduleRetractEvent: failed to allocate host mamba slot");
             }
+            hybrid_prefix_cache_->AttachHostMamba(terminal, std::move(host_slot), terminal->MambaDeviceDepthTokens());
+            pages_to_transfer.push_back(
+                {CacheTransferKind::Mamba, terminal->MambaSlotIndex(), terminal->MambaHostSlotIndex()});
```
Comment (tuanzhangCS) [inline tokenspeed-scheduler/csrc/fsm/forward_events.cpp]: Will the `mamba_host_slot` associated with a retract request be protected from eviction?

File: python/tokenspeed/runtime/cache/executor/host_executor.py [hunk 3]
```diff
@@ -173,6 +222,188 @@ class _Ack(NamedTuple):
...
+            host_slots, device_slots = src, dst
+        if not host_slots:
+            return None
+        host_indices = torch.tensor(host_slots, dtype=torch.int64)
+        device_indices = torch.tensor(device_slots, dtype=torch.int64)
+        return _TransferOp(host_indices, device_indices, op_id, is_retract)
+
+    def write(self, executor, op: _TransferOp, prepared) -> None:
+        executor._copy_mamba_slots(
+            executor.mamba_pool,
```
Comment (SimonCqk) [inline python/tokenspeed/runtime/cache/executor/host_executor.py]: QQ, in KVTransferBackend, there are `draft_pool` write/load operations, will mamba be possible  with a draft pool ?

File: tokenspeed-scheduler/csrc/scheduler/operations/forward.cpp [hunk 4]
```diff
@@ -219,9 +244,16 @@ std::optional<fsm::ScheduleDecodeFromRetractedEvent> Scheduler::scheduleDecodeFr
...
-        if (!hybrid_prefix_cache_->EnsureMambaCapacityByEvict(2, mamba_recovery_node)) {
+        std::int32_t mamba_slots_needed = 2 + static_cast<std::int32_t>(mamba_loadback_diff.size());
+        if (!hybrid_prefix_cache_->EnsureMambaCapacityByEvict(mamba_slots_needed, mamba_recovery_node)) {
             return {};
         }
+        for (TreeNode* node : mamba_loadback_diff) {
+            auto slot = hybrid_prefix_cache_->AllocateDeviceMamba();
+            if (slot == nullptr) return {};
+            hybrid_prefix_cache_->LoadBackMamba(node, std::move(slot));
+            match_result.mamba_cow_src_index = node->MambaSlotIndex();
```
Comment (SimonCqk) [inline tokenspeed-scheduler/csrc/scheduler/operations/forward.cpp]: `mamba_loadback_diff` is a vector and for now its length is alway 1, but overwrite `mamba_cow_src_index` in for-loop seems weird, should it be lifted as a vector too?

File: python/tokenspeed/runtime/layers/attention/backends/hybrid_linear_attn.py [hunk 5]
```diff
@@ -180,6 +186,20 @@ def __init__(
...
+    def make_host_pool(self, size: int) -> "SimpleMambaPool":
+        return SimpleMambaPool(
+            size=size,
+            num_mamba_layers=self.num_mamba_layers,
+            conv_state_shape=self.conv_state_shape,
+            temporal_state_shape=self.temporal_state_shape,
+            conv_dtype=self.conv_dtype,
+            ssm_dtype=self.ssm_dtype,
+            mamba_layer_ids=self.mamba_layer_ids,
+            device="cpu",
```
Comment (SimonCqk) [inline python/tokenspeed/runtime/layers/attention/backends/hybrid_linear_attn.py]: should the memory of mamba host pool be pinned?

File: python/tokenspeed/runtime/cache/executor/host_executor.py (hunk 3 above)
Comment (tuanzhangCS) [reply python/tokenspeed/runtime/cache/executor/host_executor.py]: The current draft version of Qwen3.5 does not include Mamba layers; however, I cannot confirm whether a Mamba pool will be introduced in the future or other models.

**Review by tuanzhangCS — APPROVED** [review]

**Comment by SimonCqk** [comment]
LGTM overall, since #146 was just merged, there are conflicts in some files, resolve them to merge 🚀🙌🏻

**Comment by zhyncs** [comment]
```
[ts] [2026-05-24 05:28:36,059  ATTN TP RANK 0] - INFO - Req: chatcmpl-019e5874-dee6-7301-9345-7b89d0f51c39 Finish! Accept_num_tokens_avg: 3.7803030303030303 (generation_output_processor.py:292)
[ts] [2026-05-24 05:28:36,069  ATTN TP RANK 0] - INFO - [cache_op] writeback done op_ids=[836] immediate=False (host_executor.py:401)
[ts] [2026-05-24 05:28:36,069  ATTN TP RANK 1] - INFO - [cache_op] writeback done op_ids=[836] immediate=False (host_executor.py:401)
[ts] [2026-05-24 05:28:36,069  ATTN TP RANK 2] - INFO - [cache_op] writeback done op_ids=[836] immediate=False (host_executor.py:401)
[ts] [2026-05-24 05:28:36,069  ATTN TP RANK 3] - INFO - [cache_op] writeback done op_ids=[836] immediate=False (host_executor.py:401)
```
This log is a bit noisy, @XucSh. We can wait until this CI/CD process is complete, then remove it.

**Review by zhyncs — APPROVED** [review]

**Review by SimonCqk — APPROVED** [review]

**Comment by zhyncs** [comment]
This PR appears to introduce some performance regressions in Perf CI, and they don’t look flaky. We may need to figure out what’s causing them before merging. @XucSh @SimonCqk @tuanzhangCS

**Comment by XucSh** [comment]
> This PR appears to introduce some performance regressions in Perf CI, and they don’t look flaky. We may need to figure out what’s causing them before merging. @XucSh @SimonCqk @tuanzhangCS

image
